src/test/java/com/uid2/shared/attest/AttestationTokenRetrieverTest.java

```diff
@@ -0,0 +1,105 @@
+package com.uid2.shared.attest;
+
+import com.uid2.enclave.AttestationException;
+import com.uid2.enclave.IAttestationProvider;
+import com.uid2.shared.ApplicationVersion;
+import com.uid2.shared.IClock;
+import io.vertx.core.Handler;
+import io.vertx.core.json.Json;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.net.URLConnection;
+import java.time.Instant;
+import java.util.concurrent.atomic.AtomicReference;
+import java.net.Proxy;
+import java.security.SecureRandom;
+
+import static com.uid2.shared.TestUtilites.makeInputStream;
+import static org.mockito.Mockito.*;
+
+public class AttestationTokenRetrieverTest {
```

Comment:
- Successful first attest, call attest(), expect call to httpClient, mock the response, check the token and expiry
- Attest again before expiry, move clock by 10 min, expect call to httpClient.............
- Test the response notification on success
- Test the response notification on failure
- Maybe test decryption stuff
-

Reply:
You need some tests around the scheduling - making sure that before 10 min, you don't reattest, but after 10 min you do.